cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java

```diff
@@ -0,0 +1,149 @@
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonWriter;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufOutputStream;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * {@link co.cask.http.HttpHandler} for exposing metadata of a runtime.
+ */
+@Path("v1/runtime")
+public class RuntimeHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final int MAX_MESSAGES_PER_READ = 100;
+  private static final int CHUNK_SIZE = 8192;
+  private static final Type MAP_STRING_CONSUME_REQUEST_TYPE = new TypeToken<Map<String,
+    MonitorConsumeRequest>>() { }.getType();
+
+  private final CConfiguration cConf;
+  private final MessageFetcher messageFetcher;
+
+  public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher) {
+    this.cConf = cConf;
+    this.messageFetcher = messageFetcher;
+  }
+
+  /**
+   * Gets list of topics along with offsets and limit as request and returns list of messages
+   */
+  @POST
+  @Path("/metadata")
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String data = request.content().toString(StandardCharsets.UTF_8);
+
+    Map<String, MonitorConsumeRequest> topicsToFetch = GSON.fromJson(data, MAP_STRING_CONSUME_REQUEST_TYPE);
+
+    ChunkResponder chunkResponder = responder.sendChunkStart(
+      HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE,
+                                                          "application/json; charset=utf-8"));
+
+    ByteBuf buffer = Unpooled.buffer();
+    JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
+                                                                  StandardCharsets.UTF_8));
+    jsonWriter.beginArray();
+
+    for (Map.Entry<String, MonitorConsumeRequest> entry : topicsToFetch.entrySet()) {
+      String lastMessageId = null;
+      int limit = entry.getValue().getLimit();
+      int batchSize = Math.min(MAX_MESSAGES_PER_READ, limit);
+      List<MonitorMessage> messages = Lists.newArrayListWithCapacity(MAX_MESSAGES_PER_READ);
+
+      // read first batch of messages
+      int eventsRead = readMessages(messages, cConf.get(entry.getKey()), batchSize, lastMessageId);
```

Comment:
Why need to read messages into at list and then iterate again to json encode one by one? You can just read from the `Iterator` returned from the `MessageFetcher.fetch` call and do the encode on the fly without buffering.